### PR TITLE
prevent deadlocking when `QueryEvents()` is not read fully

### PR DESCRIPTION
**Pull request by fiatjaf**
this could happen in relay queries that exit before EOSE.

the channel inside `QueryEvents()` of every backend will just halt since no one is reading from it.

### CODE REVIEW

**Comment by fiatjaf** [comment]
please take a look @mattn, is this the wrong approach?

**Comment by fiatjaf** [comment]
Actually this has happened in the real world not because of a broken connection, but because @pablof7z was calling `.QueryEvents()` manually and then not reading the returned channel until completion.

I still think it's a bug on this library to not handle that case correctly, but maybe we shouldn't fix and we should just say "beware: channels must be read until completion"?

File: elasticsearch/query.go [hunk 1]
```diff
@@ -123,9 +123,15 @@ func (ess *ElasticsearchStorage) QueryEvents(ctx context.Context, filter nostr.F
...
+			go func() {
+				for _, evt := range evts {
+					select {
+					case ch <- evt:
+					case <-ctx.Done():
+						return
+					}
+				}
+			}()
 			close(ch)
```
Comment (mattn) [inline elasticsearch/query.go:135]: This seems to be occur panic with sending to closed channel.

Comment (mattn) [reply elasticsearch/query.go:135]: can you move this close into goroutine with defer close(ch) ?

Comment (fiatjaf) [reply elasticsearch/query.go]: Moved.

**Comment by mattn** [comment]
LGTM

**Comment by mattn** [comment]
Thank you

**Pull request merged**